File: android/src/main/java/com/margelo/nitro/simpleimagecompressor/SimpleImageCompressor.kt
package com.margelo.nitro.simpleimagecompressor

import android.content.Context
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise


@DoNotStrip
class SimpleImageCompressor : HybridSimpleImageCompressorSpec() {

  private fun getContext(): Context {
    val context =
      NitroModules.applicationContext ?: throw Exception("Nitromodule.applicationContext is null")
    return context
  }

  override fun getFileSize(uri: String): Promise<Double> {
    return Promise.async {

      val sourceUri = getCleanUri(uri)
      val context = getContext()
      fetchFileSizeByUri(context, sourceUri).toDouble()
    }
  }

  override fun compressImage(uri: String, options: CompressOptions): Promise<CompressedResult> {
    return Promise.async {
      val context = getContext()

      SimpleCompressorService.compress(
        context,
        uri,
        options.quality,
        options.maxWidth?.toInt(),
        options.maxHeight?.toInt(),
        options.format,
        options.enablePhysicalRotation == true

      )
    }
  }
}

File: ios/SimpleImageCompressor.swift
import Foundation
import NitroModules

class SimpleImageCompressor: HybridSimpleImageCompressorSpec {
  func getFileSize(uri: String) throws -> NitroModules.Promise<Double> {
    return Promise.async {
      let sourceUrl = try FileUtils.getCleanUri(uriString: uri)
      let size = try FileUtils.fetchFileSize(fileUrl: sourceUrl)
      return Double(size)
    }
  }

  var memorySize: Int { return 0 }

  public func compressImage(uri: String, options: CompressOptions) throws
    -> Promise<CompressedResult>
  {
    return Promise.async {

      let sourceUrl = try FileUtils.getCleanUri(uriString: uri)

      let quality = options.quality
      let maxWidth = options.maxWidth.map { Int($0) }
      let maxHeight = options.maxHeight.map { Int($0) }

      let imageFormat: ImageFormat
      switch options.format {
      case .png:
        imageFormat = .png
      case .webp:
        imageFormat = .webp
      case .webpLossless:
        imageFormat = .webpLossless
      default:
        imageFormat = .jpg
      }

      let compressedResult = try ImageCompressorService.compress(
        sourceUrl: sourceUrl,
        quality: quality,
        maxWidth: maxWidth,
        maxHeight: maxHeight,
        imageFormat: imageFormat,
        enablePhysicalRotation: options.enablePhysicalRotation ?? false
      )

      let outputImageFormat: OutputCompressedFormat
      switch compressedResult.format {
      case .png:
        outputImageFormat = .png
      case .jpg:
        outputImageFormat = .jpg
      case .webp:
        outputImageFormat = .webp
      case .webpLossless:
        outputImageFormat = .webpLossless
      }

      return CompressedResult(
        uri: compressedResult.uri.absoluteString,
        width: Double(compressedResult.width),
        height: Double(compressedResult.height),
        format: outputImageFormat,
        fileSize: Double(compressedResult.fileSize),
        originalFileSize: Double(compressedResult.originalFileSize)
      )
    }
  }
}
